### PR TITLE
vdirsyncerStable: init at 0.16.7

### LINKED ISSUE #58563: vdirsyncer fails to sync
## Issue description
`vdirsyncer sync` fails with the vdirsyncer version from nixos-unstable. it works if I install vdirsyncer with pip.

```
$ vdirsyncer -vdebug sync 
debug: Using 1 maximal workers.
debug: Fetching value for password.fetch with command strategy.
Syncing contacts/32
error: Unknown error occurred for contacts/32: Server did not return etag.
error: Use `-vdebug` to see the full traceback.
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/cli/tasks.py", line 69, in sync_collection
debug:     partial_sync=pair.partial_sync
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/sync/__init__.py", line 143, in sync
debug:     b_nonempty = b_info.prepare_new_status()
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/sync/__init__.py", line 60, in prepare_new_status
debug:     if prefetch else ()):
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/storage/base.py", line 149, in get_multi
debug:     item, etag = self.get(href)
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/storage/_rust.py", line 42, in get
debug:     native.check_error(e)
debug:   File "/nix/store/y9qgx26b2w2lzqhqb7sd9cb6cz8dj7pa-vdirsyncer-unstable-2018-08-05/lib/python3.7/site-packages/vdirsyncer/native.py", line 35, in check_error
debug:     errors.check_exception(e[0])
debug:   File "/nix/store/4xg6a6pqm9v3b1vfmjncqy0w18iqfrhj-python3.7-shippai-0.3.2/lib/python3.7/site-packages/shippai/__init__.py", line 150, in check_exception
debug:     raise exc
error: 1 out of 2 tasks failed.
```

## Technical details

Output of `nix-shell -p nix-info --run "nix-info -m"`:
```
- system: `"x86_64-linux"`
 - host os: `Linux 4.19.30, NixOS, 19.09pre173349.07b42ccf2de (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"nixos-19.09pre173349.07b42ccf2de"`
 - channels(****): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by loewenheim**
###### Motivation for this change
Adds a package for the latest stable release of `vdirsyncer`.
fixes #58563

###### Things done
- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{ stdenv, python3Packages, fetchFromGitHub, fetchpatch, glibcLocales, pkgconfig, openssl, Security }:
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]:
```suggestion
{ lib, python3Packages, fetchpatch }:
```

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+# Packaging documentation at:
+# https://github.com/untitaker/vdirsyncer/blob/master/docs/packaging.rst
+python3Packages.buildPythonApplication rec {
+  version = "0.16.7";
+  pname = "vdirsyncer";
+  name = "${pname}-${version}";
+
+  src = fetchFromGitHub {
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]: Please use `fetchPypi`.

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
+{ stdenv, python3Packages, fetchFromGitHub, fetchpatch, glibcLocales, pkgconfig, openssl, Security }:
+
+# Packaging documentation at:
+# https://github.com/untitaker/vdirsyncer/blob/master/docs/packaging.rst
+python3Packages.buildPythonApplication rec {
+  version = "0.16.7";
+  pname = "vdirsyncer";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]:
```suggestion
```

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 4]
```diff
@@ -0,0 +1,67 @@
+{ stdenv, python3Packages, fetchFromGitHub, fetchpatch, glibcLocales, pkgconfig, openssl, Security }:
+
+# Packaging documentation at:
+# https://github.com/untitaker/vdirsyncer/blob/master/docs/packaging.rst
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]: That's not the correct URL for this version.

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    requests
+    requests_oauthlib # required for google oauth sync
+    atomicwrites
+    milksnake
+    shippai
+  ];
+
+  buildInputs = with python3Packages; [ setuptools_scm ];
+
+  checkInputs = with python3Packages; [ hypothesis pytest pytest-localserver pytest-subtesthack ] ++ [ glibcLocales ];
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]:
```suggestion
  checkInputs = with python3Packages; [ hypothesis pytest pytest-localserver pytest-subtesthack ];
```

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    click click-log click-threading
+    requests_toolbelt
+    requests
+    requests_oauthlib # required for google oauth sync
+    atomicwrites
+    milksnake
+    shippai
+  ];
+
+  buildInputs = with python3Packages; [ setuptools_scm ];
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]:
```suggestion
  nativeBuildInputs = with python3Packages; [ setuptools_scm ];
```

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    substituteInPlace tests/unit/test_repair.py \
+      --replace $'@settings(perform_health_check=False)  # Using the random module for UIDs\n' ""
+
+    # for setuptools_scm:
+    echo 'Version: ${version}' >PKG-INFO
+
+    sed -i 's/spec.add_external_build(cmd=cmd/spec.add_external_build(cmd="true"/g' setup.py
+  '';
+
+  LC_ALL = "en_US.utf8";
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]:
```suggestion
```

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+  checkPhase = ''
+    rm -rf vdirsyncer
+    export PYTHONPATH=$out/${python3Packages.python.sitePackages}:$PYTHONPATH
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]: This line is probably not necessary.

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    make DETERMINISTIC_TESTS=true test
+  '';
+
+  meta = with stdenv.lib; {
+    homepage = https://github.com/pimutils/vdirsyncer;
+    description = "Synchronize calendars and contacts";
+    maintainers = with maintainers; [ matthiasbeyer ];
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]: ??

File: pkgs/tools/misc/vdirsyncerStable/default.nix [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    # Invalid argument: 'perform_health_check' is not a valid setting
+    substituteInPlace tests/conftest.py \
+      --replace "perform_health_check=False" ""
+    substituteInPlace tests/unit/test_repair.py \
+      --replace $'@settings(perform_health_check=False)  # Using the random module for UIDs\n' ""
+
+    # for setuptools_scm:
+    echo 'Version: ${version}' >PKG-INFO
+
+    sed -i 's/spec.add_external_build(cmd=cmd/spec.add_external_build(cmd="true"/g' setup.py
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncerStable/default.nix]: Please check whether these are still necessary.

**Review by dotlambda — CHANGES_REQUESTED** [review]
I'd prefer adding a `stable.nix` in the vdirsyncer directory.

**Comment by loewenheim** [comment]
Thank you for the review. I’ll make changes as soon as I’m able.

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 11]
```diff
@@ -0,0 +1,55 @@
+{ lib, python3Packages, fetchpatch }:
+
+# Packaging documentation at:
+# https://github.com/pimutils/vdirsyncer/blob/dcf5f701b7b5c21a8f4e8c80243db3e0baff1313/docs/packaging.rst
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncer/stable.nix]:
```suggestion
# https://github.com/pimutils/vdirsyncer/blob/0.16.7/docs/packaging.rst
```

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    sed -i 's/spec.add_external_build(cmd=cmd/spec.add_external_build(cmd="true"/g' setup.py
+  '';
+
+  meta = with lib; {
+    homepage = https://github.com/pimutils/vdirsyncer;
+    description = "Synchronize calendars and contacts";
+    platforms = platforms.all;
+    license = licenses.mit;
+  };
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncer/stable.nix]: According to https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#submitting-changes, a maintainer is required.

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+    # for setuptools_scm:
+    echo 'Version: ${version}' >PKG-INFO
+
+    sed -i 's/spec.add_external_build(cmd=cmd/spec.add_external_build(cmd="true"/g' setup.py
+  '';
+
+  meta = with lib; {
+    homepage = https://github.com/pimutils/vdirsyncer;
+    description = "Synchronize calendars and contacts";
+    platforms = platforms.all;
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncer/stable.nix]:
```suggestion
```
That's set automatically by `buildPythonApplication`.

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  propagatedBuildInputs = with python3Packages; [
+    click click-log click-threading
+    requests_toolbelt
+    requests
+    requests_oauthlib # required for google oauth sync
+    atomicwrites
+    milksnake
+    shippai
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncer/stable.nix]: I can't find the last two in https://github.com/pimutils/vdirsyncer/blob/0.16.7/setup.py.

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    # Invalid argument: 'perform_health_check' is not a valid setting
+    substituteInPlace tests/conftest.py \
+      --replace "perform_health_check=False" ""
+    substituteInPlace tests/unit/test_repair.py \
+      --replace $'@settings(perform_health_check=False)  # Using the random module for UIDs\n' ""
+
+    # for setuptools_scm:
+    echo 'Version: ${version}' >PKG-INFO
+
+    sed -i 's/spec.add_external_build(cmd=cmd/spec.add_external_build(cmd="true"/g' setup.py
```
Comment (dotlambda) [inline pkgs/tools/misc/vdirsyncer/stable.nix]: Most of these lines (except for the first three) should be removed.

**Review by dotlambda — CHANGES_REQUESTED** [review]
Please squash your new commits into the existing one.

File: pkgs/tools/misc/vdirsyncerStable/default.nix (hunk 8 above)
Comment (dotlambda) [reply pkgs/tools/misc/vdirsyncerStable/default.nix]: Why did you just remove the entire `checkPhase`? Please keep the third line.

**Comment by dotlambda** [comment]
@gebner @matthiasbeyer Many of my comments also apply to the `vdirsyncer` expression. It would be nice if you could improve it.

**Comment by gebner** [comment]
@dotlambda Thanks for looking over it, I've made the changes to the default vdirsyncer expression.

File: pkgs/tools/misc/vdirsyncer/stable.nix (hunk 12 above)
Comment (loewenheim) [reply pkgs/tools/misc/vdirsyncer/stable.nix]: In that case, should I just add myself to the maintainer list?

Comment (dotlambda) [reply pkgs/tools/misc/vdirsyncer/stable.nix]: Yes, please do so in a separate commit preceding the existing one.

File: pkgs/tools/misc/vdirsyncer/stable.nix [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+  checkPhase = ''
+    make DETERMINISTIC_TESTS=true test
+  '';
+
+  meta = with lib; {
+    homepage = https://github.com/pimutils/vdirsyncer;
+    description = "Synchronize calendars and contacts";
+    license = licenses.mit;
+    maintainers = [ loewenheim ];
```
Comment (gebner) [inline pkgs/tools/misc/vdirsyncer/stable.nix]:
```suggestion
    maintainers = with maintainers; [ loewenheim ];
```

The maintainers are in `lib.maintainers`.  This error here causes evaluation to fail, as you can see from the failing @GrahamcOfBorg check below.

Comment (loewenheim) [reply pkgs/tools/misc/vdirsyncer/stable.nix]: Thank you. Is it possible to run the check on my machine so I can see if it works?

Comment (gebner) [reply pkgs/tools/misc/vdirsyncer/stable.nix]:
```shell
env VERBOSE=1 ./maintainers/scripts/eval-release.sh
```

**Comment by loewenheim** [comment]
Thank you again @dotlambda @gebner for your guidance.

**Comment by matthiasbeyer** [comment]
Awesome! Now ... may I ask what "stable" means in the context of vdirsyncer? Because I really don't see how a "0.x.y" package can be "stable".

Or, phrased in another way: How do you want to update the new "stable" package? Only from `0.16.x -> 0.16.y`?

**Comment by dotlambda** [comment]
As you can see [here](https://github.com/pimutils/vdirsyncer/releases), the 0.17.0 releases are still considered alpha. Hence I suggest keeping the vdirsyncerStable expression until these are considered stable, then making it an alias for vdirsyncer.

**Pull request merged**